In [6]:
pip install --user -U nltk

distutils: /Users/juliette/Library/Python/3.9/lib/python/site-packages
sysconfig: /Users/juliette/Library/Python/3.9/lib/python3.9/site-packages
distutils: /Users/juliette/Library/Python/3.9/lib/python/site-packages
sysconfig: /Users/juliette/Library/Python/3.9/lib/python3.9/site-packages
user = True
home = None
root = None
prefix = None
You should consider upgrading via the '/usr/local/Cellar/jupyterlab/3.3.2/libexec/bin/python3.9 -m pip install --upgrade pip' command.
Note: you may need to restart the kernel to use updated packages.


In [4]:
pip install gensim

     |████████████████████████████████| 24.0 MB 3.7 MB/s            
     |████████████████████████████████| 56 kB 3.1 MB/s            
You should consider upgrading via the '/usr/local/Cellar/jupyterlab/3.3.2/libexec/bin/python3.9 -m pip install --upgrade pip' command.
Note: you may need to restart the kernel to use updated packages.


In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
#for text pre-processing
import re, string
import nltk
from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
from nltk.stem import SnowballStemmer
from nltk.corpus import wordnet
from nltk.stem import WordNetLemmatizer
nltk.download('punkt')
nltk.download('averaged_perceptron_tagger')
nltk.download('wordnet')
#for model-building
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import classification_report, f1_score, accuracy_score, confusion_matrix
from sklearn.metrics import roc_curve, auc, roc_auc_score
# bag of words
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.feature_extraction.text import CountVectorizer
#for word embedding
import gensim
from gensim.models import Word2Vec

[nltk_data] Downloading package punkt to /Users/juliette/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /Users/juliette/nltk_data...
[nltk_data]   Package averaged_perceptron_tagger is already up-to-
[nltk_data]       date!
[nltk_data] Downloading package wordnet to
[nltk_data]     /Users/juliette/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


In [34]:
df_train = pd.read_json("train.jsonl",lines=True)
df_test = pd.read_json("test.jsonl",lines=True)
df_dev = pd.read_json("dev.jsonl",lines=True)

In [27]:
df_train.head()

id            img  label  \
0  42953  img/42953.png      0   
1  23058  img/23058.png      0   
2  13894  img/13894.png      0   
3  37408  img/37408.png      0   
4  82403  img/82403.png      0   

                                                text  
0   its their character not their color that matters  
1  don't be afraid to love again everyone is not ...  
2                           putting bows on your pet  
3  i love everything and everybody! except for sq...  
4  everybody loves chocolate chip cookies, even h...

# EDA

In [5]:
#Distribution : 
x=df_train['label'].value_counts()
print(x)

0    5450
1    3050
Name: label, dtype: int64


Il y a 5450 memes non haineux et 3050 memes haineux

In [6]:
# Vérification des NA
df_train.isna().sum()

id       0
img      0
label    0
text     0
dtype: int64

Il n'y a aucune valeur manquante

In [43]:
# Nombre de mots 
df_train_2 = df_train.copy()
df_train_2['word_count'] = df_train_2['text'].apply(lambda x: len(str(x).split()))
print(df_train_2[df_train_2['label']==1]['word_count'].mean()) # memes haineux
print(df_train_2[df_train_2['label']==0]['word_count'].mean()) # memes non haineux

12.79344262295082
11.154495412844037


Il y a environ le même nombre de mots entre les deux types de même (en moyenne un de plus pour les mêmes haineux)

In [44]:
# Nombre de lettres
df_train_2['char_count'] = df_train_2['text'].apply(lambda x: len(str(x)))
print(df_train_2[df_train_2['label']==1]['char_count'].mean()) #memes haineux
print(df_train_2[df_train_2['label']==0]['char_count'].mean()) #memes non haineux

68.48360655737704
58.489724770642205


Il y a environ une dizaine de caractères en plus dans les mêmes haineux

In [45]:
df_train_2.head()

id            img  label  \
0  42953  img/42953.png      0   
1  23058  img/23058.png      0   
2  13894  img/13894.png      0   
3  37408  img/37408.png      0   
4  82403  img/82403.png      0   

                                                text  word_count  char_count  
0   its their character not their color that matters           8          48  
1  don't be afraid to love again everyone is not ...          12          58  
2                           putting bows on your pet           5          24  
3  i love everything and everybody! except for sq...          11          70  
4  everybody loves chocolate chip cookies, even h...           7          51

# Preparation de la base avant la classification

In [9]:
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to
[nltk_data]     /Users/juliette/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

In [10]:
#On enlève la ponctuatoin et les strips
def preprocess(text):
    text = text.lower() 
    text=text.strip()  #suppression des caractères de tête et de queue
    text=re.compile('<.*?>').sub('', text) 
    text = re.compile('[%s]' % re.escape(string.punctuation)).sub(' ', text)  
    text = re.sub('\s+', ' ', text)  
    text = re.sub(r'\[[0-9]*\]',' ',text) 
    text=re.sub(r'[^\w\s]', '', str(text).lower().strip())
    text = re.sub(r'\d',' ',text) 
    text = re.sub(r'\s+',' ',text) 
    return text

 
# STOPWORD REMOVAL : on enlève les mots inutiles "the" etc.
def stopword(string):
    a= [i for i in string.split() if i not in stopwords.words('english')]
    return ' '.join(a)

#LEMMATIZATION 
# On initialise le lemmatizer
wl = WordNetLemmatizer()
 
# This is a helper function to map NTLK position tags

#WordNetLemmatizer a besoin des balises Pos pour comprendre si le mot est un nom, 
#un verbe, un adjectif, etc. Par défaut, elle est définie sur Noun

def get_wordnet_pos(tag):
    if tag.startswith('J'):
        return wordnet.ADJ
    elif tag.startswith('V'):
        return wordnet.VERB
    elif tag.startswith('N'):
        return wordnet.NOUN
    elif tag.startswith('R'):
        return wordnet.ADV
    else:
        return wordnet.NOUN

# Tokenize 
def lemmatizer(string):
    word_pos_tags = nltk.pos_tag(word_tokenize(string)) # Get position tags
    a=[wl.lemmatize(tag[0], get_wordnet_pos(tag[1])) for idx, tag in enumerate(word_pos_tags)] # Map the position tag and lemmatize the word/token
    return " ".join(a)



In [12]:
import nltk
nltk.download('omw-1.4')

[nltk_data] Downloading package omw-1.4 to
[nltk_data]     /Users/juliette/nltk_data...


True

In [47]:
def finalpreprocess(string):
    return lemmatizer(stopword(preprocess(string)))
df_train['clean_text'] = df_train['text'].apply(lambda x: finalpreprocess(x))
df_dev['clean_text'] = df_dev['text'].apply(lambda x: finalpreprocess(x))
df_train.head()

id            img  label  \
0  42953  img/42953.png      0   
1  23058  img/23058.png      0   
2  13894  img/13894.png      0   
3  37408  img/37408.png      0   
4  82403  img/82403.png      0   

                                                text  \
0   its their character not their color that matters   
1  don't be afraid to love again everyone is not ...   
2                           putting bows on your pet   
3  i love everything and everybody! except for sq...   
4  everybody loves chocolate chip cookies, even h...   

                                          clean_text  
0                             character color matter  
1                       afraid love everyone like ex  
2                                        put bow pet  
3  love everything everybody except squirrel hate...  
4    everybody love chocolate chip cooky even hitler

In [48]:
df_dev.head()

id            img  label  \
0   8291  img/08291.png      1   
1  46971  img/46971.png      1   
2   3745  img/03745.png      1   
3  83745  img/83745.png      1   
4  80243  img/80243.png      1   

                                                text  \
0              white people is this a shooting range   
1                              bravery at its finest   
2  your order comes to $37.50 and your white priv...   
3  it is time.. to send these parasites back to t...   
4                             mississippi wind chime   

                                         clean_text  
0                          white people shoot range  
1                                      bravery fine  
2  order come white privilege discount brings total  
3                    time send parasite back desert  
4                            mississippi wind chime

In [60]:
# Mise en vecteur (word embedding)

#Création du train et du test (on test sur le dev car on a pas les labels sur le test)

X_train, y_train = (df_train["clean_text"],df_train["label"])
X_test, y_test = (df_dev["clean_text"],df_dev["label"])

#X contient les mots et y contient les labels correspondants



In [61]:
X_train.head()

0                               character color matter
1                         afraid love everyone like ex
2                                          put bow pet
3    love everything everybody except squirrel hate...
4      everybody love chocolate chip cooky even hitler
Name: clean_text, dtype: object

In [63]:
y_train.head()

0    0
1    0
2    0
3    0
4    0
Name: label, dtype: int64

In [64]:
X_test.head()

0                            white people shoot range
1                                        bravery fine
2    order come white privilege discount brings total
3                      time send parasite back desert
4                              mississippi wind chime
Name: clean_text, dtype: object

In [65]:
y_test.head()

0    1
1    1
2    1
3    1
4    1
Name: label, dtype: int64

In [66]:
#Tf-Idf (méthode pour mathématiser le texte)
tfidf_vectorizer = TfidfVectorizer(use_idf=True)
X_train_vectors_tfidf = tfidf_vectorizer.fit_transform(X_train) # on obtient des matrices
X_test_vectors_tfidf = tfidf_vectorizer.transform(X_test)

# Classification

## Régression linéaire

In [67]:
#FITTING THE CLASSIFICATION MODEL using Logistic Regression(tf-idf)
lr_tfidf=LogisticRegression(solver = 'liblinear', C=10, penalty = 'l2')
lr_tfidf.fit(X_train_vectors_tfidf, y_train)  

#Predict y value for test dataset
y_predict = lr_tfidf.predict(X_test_vectors_tfidf)
y_prob = lr_tfidf.predict_proba(X_test_vectors_tfidf)[:,1]
print(classification_report(y_test,y_predict))
print('Confusion Matrix:',confusion_matrix(y_test, y_predict))
 
fpr, tpr, thresholds = roc_curve(y_test, y_prob)
roc_auc = auc(fpr, tpr)
print('AUC:', roc_auc)

              precision    recall  f1-score   support

           0       0.51      0.81      0.63       250
           1       0.55      0.23      0.32       250

    accuracy                           0.52       500
   macro avg       0.53      0.52      0.48       500
weighted avg       0.53      0.52      0.48       500

Confusion Matrix: [[203  47]
 [193  57]]
AUC: 0.566544


*Interprétation* : 

   *La précision* : 
   
Sur l'ensemble des memes que le modèle à prédit comme non haineux (label 1) seuls 55% le sont réellement.
Sur l'ensemble des memes que le modèle à prédit comme haineux (label 0), 51% le sont réellement.

      

*L'accuracy* : 

La proportion de mèmes correctement prédis est de 0.52. Ce n'est pas un bon résultat. 

CCl : Mauvaise prédiction pour les memes non haineux. 

## Naive Bayes Classifier Algorithm

In [68]:
from sklearn import model_selection, naive_bayes, svm
from sklearn.metrics import accuracy_score


In [69]:
# fit the training dataset on the NB classifier

Naive = naive_bayes.MultinomialNB()
Naive.fit(X_train_vectors_tfidf,y_train)

# predict the labels on validation dataset

predictions_NB = Naive.predict(X_test_vectors_tfidf)


#classification report
print(classification_report(y_test,predictions_NB))

# Accuracy
print("Naive Bayes Accuracy Score -> ",accuracy_score(predictions_NB, y_test))

              precision    recall  f1-score   support

           0       0.51      0.94      0.66       250
           1       0.59      0.08      0.14       250

    accuracy                           0.51       500
   macro avg       0.55      0.51      0.40       500
weighted avg       0.55      0.51      0.40       500

Naive Bayes Accuracy Score ->  0.512


On retrouve une accuracy de 0.51. Ce n'est pas un bon résulat. 

## SVM — Support Vector Machine

In [70]:
# fit the training dataset on the classifier

SVM = svm.SVC(C=1.0, kernel='linear', degree=3, gamma='auto')
SVM.fit(X_train_vectors_tfidf,y_train)

# predict the labels on validation dataset

predictions_SVM = SVM.predict(X_test_vectors_tfidf)

#classification report
print(classification_report(y_test,predictions_SVM))

# Use accuracy_score function to get the accuracy

print("SVM Accuracy Score -> ",accuracy_score(predictions_SVM, y_test))

              precision    recall  f1-score   support

           0       0.52      0.89      0.66       250
           1       0.63      0.19      0.29       250

    accuracy                           0.54       500
   macro avg       0.58      0.54      0.48       500
weighted avg       0.58      0.54      0.48       500

SVM Accuracy Score ->  0.54


On obtient une meilleure accuracy qu'avec les modèles précédents: 0.54. Cependant, ce résultat n'est toujours pas satisfaisant. 

On en déduit que l'analyse textuelle seule n'est pas suffisante pour catégoriser les mèmes en haineux ou non haineux. 